### ISSUE TITLE
Initial settings.ini & bat/sh files automation?

### ISSUE THREAD

**Issue by TiddlyTweeter** [Marxsal/polly#6]
Thinking (though I have zero knowledge of Linux), that I could modify **do-settings.ps1** so that on first-run it creates ...

1. settings.ini
      and...
2. the three basic "batch" files--according to PLATFORM. So "bat" for Win and "sh" for Linux?

**Comment by TiddlyTweeter** [comment]
But would clicking  **polly.sh** in Linux work the same way as on Win?

**Comment by TiddlyTweeter** [comment]
Thinking this on a bit ... I don't like idea we'd have to maintain three different "**settings booter"**. I'd hope we can produce _(fingers crossed)_ a reliable cross-platform system. So, right now I'm thinking a "universal" release might look like this ...

```
polly.sh
polly.bat
polly.ps1

.\lib (to include needed libs and all platform dependent scripts) 
```
According to platform appropriate **settings.ini** and **once.'platform'** and **auto.'platform'** would be created on first run. (this assumes ".sh" won't run on Win & ".bat" won't run on Linux or Mac _[it might???]_)

**Comment by Marxsal** [comment]
A .bat file can run in lin*x, but it has to be marked as executable (e.g. chmod 764 polly2.bat). But for some reason the command to launch PS isn't "powershell", but "pwsh". and you can't use the backslash (.\pwsh) but instead the forward slash (./pwsh).

So it seems like the main way to distribute polly is via two distributions, one for un*x, and one for Windows. This might not be so hard. There can be two branches for distribution. One for un*x, and the other for Windows, both based on master with the only difference being the initial batch files.

**Comment by TiddlyTweeter** [comment]
Just FYI I think it would be relatively easy in do-settings.ps1 to add PS version and platform detection and auto generate appropriate .bat / .sh / .ini files, either by copying templates or by dynamically building the needed files. 

_Q: Would you like me to work on that?_

**Comment by Marxsal** [comment]
@TiddlyTweeter That would be neat. But how would do-settings.ps1 get launched in the first place?
If you generate .sh files on the fly, you will probably also need to modify them with a command like:

chmod u+x myfile.sh

This adds the executable permission at the user level. chmod is on all Linux platforms, afaik. Probably Mac too?

Hopefully that's possible from script. I guess we'll find out.

**Comment by TiddlyTweeter** [comment]
> @Marxsal That would be neat. But how would do-settings.ps1 get launched in the first place?

AIDE-MEMOIRE: _notes on suggested process based on work to date ..._ On first install the app. dir contains only ...

```
setup.bat
setup.sh
polly.ps1
./lib 
```

1.  The "setup" files simply launch polly.ps1 with the appropriate version of PS. The Windows .bat to contain detection whether "pwsh" is present and runs using it, else runs with powershell.

2. polly.ps 1 detects that settings.ini is missing, and invokes ./lib/do-settings.ps1.

3. do-settings.ps1 creates basic bat/sh for the run environment (determined by PS version + IsPlatform) ...

4. ... applies files permissions to them where needed

5. ... then deletes "setup.*" files.

6. ... then initiates settings.ini creation as we have it already. 

IF at this point the user quits its ok. Next time they launch Polly it will not need step 1, but otherwise work the same just using any of the created "batch" files that are now "environment specific".

**Comment by TiddlyTweeter** [comment]
Hi @Marxsal could you test for me what happens in a Linux PS6 console when you enter ...

"`[Environment]::EDITOR`"

... and any of these on a file ...

`chmod  u+x  test.sh`

`chmod  u+x  *.sh` (will take wildcards?)

`Start-Process -FilePath "chmod"  -ArgumentList "u+x  test.sh" -WindowStyle "minimized"` (may open new console? Won't take wildcards?)

**Comment by Marxsal** [comment]
> `[Environment]::EDITOR
`

Returns nothing. I suspect that it is one of those sometimes-used variables. You might have to check for the most popular/likely editors (which geany, which nano, which vim, which vi, which emacs) and then, if all else fails, print a message to the user telling them what they need to do.

> `chmod u+x test.sh`
> 
> `chmod u+x *.sh` (will take wildcards?)
> 
> `Start-Process -FilePath "chmod" -ArgumentList "u+x test.sh" -NoNewWindow`

These all work, and return you to the PS6 terminal.

**Comment by TiddlyTweeter** [comment]
`[Environment]::EDITOR`

> Returns nothing. I suspect that it is one of those sometimes-used variables. You might have to check for the most popular/likely editors (which geany, which nano, which vim, which vi, which emacs) and then, if all else fails, print a message to the user telling them what they need to do.

@Marxsal, one further question about Linux. Does it support launch by "file type" like Windows, e.g. on Win you can open a text editor simply via ` Start-Process "file.txt" ` since the type has a defined (registry entry) association to an ".exe".

**Comment by TiddlyTweeter** [comment]
I updated this issue on what a "release" might look like at https://github.com/Marxsal/polly/issues/6#issuecomment-522207438

**Comment by Marxsal** [comment]
> Does it support launch by "file type" like Windows, e.g. on Win you can open a text editor simply via `Start-Process "file.txt"` since the type has a defined (registry entry) association to an ".exe".

I would say "not reliably". I tried it and although it opened the file in a terminal window, the window was "broken", with prompts and text appearing at the wrong locations on the screen and keystrokes obscured.

What apparently works on Debian/Ubuntu distributions is to type 

`editor temp.txt
`
this will invoke the default editor. But it may not apply to all distributions.

**Comment by TiddlyTweeter** [comment]
Much as I value theoretical idea of Linux it sounds like a nightmare to work with.

**Comment by TiddlyTweeter** [comment]
TBH, I think it may be easier to just create settings.ini in Powershell??

**Comment by Marxsal** [comment]
> 
> 
> TBH, I think it may be easier to just create settings.ini in Powershell??

I thought that was what your were planning on doing ??

**Comment by TiddlyTweeter** [comment]
> I thought that was what your were planning on doing ??

Yeah. Sort of. But not because of lack of a text editor you could rely on. 

Now its more of an imperative to avoid hassle.
